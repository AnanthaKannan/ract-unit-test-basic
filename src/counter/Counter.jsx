import React, {useState, useEffect} from 'react'

export default function Counter() {

    const [counterValue, setCounterValue] = useState(0);
    const [inputValue, setInputValue] = useState(1);
    const [className, setClassName] = useState("")

    const counter = (status) => {
        if(status === 'add'){
            setCounterValue(inputValue + counterValue)
        }
        else{
            setCounterValue(counterValue - inputValue)
        }
    }

    useEffect(() => {
        if(counterValue >= 100){
            setClassName('bg-success')
        }
        else if(counterValue <= -100){
            setClassName('bg-danger')
        }
        else{
            setClassName('')
        }
    },[counterValue])

    return (
        <div>
            <h2 data-testid='header'>My Counter</h2>
            <h3 
            className={className}
            data-testid='counter'
            >{counterValue}</h3>

            <button 
            data-testid='add-btn'
            onClick={() => counter('add')}
            >+</button>

            <input 
            className='App'
            type="number" 
            data-testid='input' 
            value={inputValue}
            onChange={(e) => setInputValue(Number(e.target.value))} />

            <button 
            data-testid='sub-btn' 
            onClick={() => counter('sub')}>-</button>

        </div>
    )
}
