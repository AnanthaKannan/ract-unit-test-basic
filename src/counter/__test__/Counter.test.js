import React from "react"
import Counter from "../Counter"
import { render, fireEvent } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect"
import renderer from 'react-test-renderer';

let getByTestId, headerEl, counterEl, inputEl, subBtnEl, addBtnEl;
beforeEach(() => {
    const component = render(<Counter />);
    getByTestId = component.getByTestId;
    headerEl = getByTestId('header');
    counterEl = getByTestId('counter');
    inputEl = getByTestId('input');
    subBtnEl = getByTestId('sub-btn');
    addBtnEl = getByTestId('add-btn');
});

test("header renders with correct text", () => {
    expect(headerEl.textContent).toBe("My Counter");
});

test("counter initially start with text of 0", () => {
    expect(counterEl.textContent).toBe("0");
});

test("input contains initial value of 1", () => {
    expect(inputEl.value).toBe("1");
});

test("add button render with +", () => {
    expect(addBtnEl.textContent).toBe("+");
});

test("subtract button render with -", () => {
    expect(subBtnEl.textContent).toBe("-");
});

test("change value of input works correctly", () => {
    expect(inputEl.value).toBe("1")
    fireEvent.change(inputEl, {
        target: {
            value: "5"
        }
    });
    expect(inputEl.value).toBe("5")
});

test("click on plus btn adds 1 to counter", () => {
    expect(counterEl.textContent).toBe("0");
    fireEvent.click(addBtnEl);
    expect(counterEl.textContent).toBe("1");
});

test("click on subtract btn sub 1 to counter", () => {
    expect(counterEl.textContent).toBe("0");
    
    fireEvent.click(subBtnEl);
    
    expect(counterEl.textContent).toBe("-1");
});

test("changing input value then clicking on add btn works correctly", () => {
    const addBtnEl = getByTestId('add-btn');
    const counterEl = getByTestId('counter');
    const inputEl = getByTestId('input');

    fireEvent.change(inputEl, {
        target: {
            value: "5"
        }
    });

    expect(counterEl.textContent).toBe("0");
    fireEvent.click(addBtnEl);
    expect(counterEl.textContent).toBe("5");
});

test("changing input value then clicking on sub btn works correctly", () => {

    fireEvent.change(inputEl, {
        target: {
            value: "5"
        }
    });
    expect(counterEl.textContent).toBe("0");
    fireEvent.click(subBtnEl);
    expect(counterEl.textContent).toBe("-5");
});

test("adding and then subtracting leads to the correct counter number", () => {

    fireEvent.change(inputEl, {
        target: {
            value: "10"
        }
    });

    expect(counterEl.textContent).toBe("0");
    fireEvent.click(addBtnEl);
    fireEvent.click(addBtnEl);
    fireEvent.click(addBtnEl);
    fireEvent.click(addBtnEl);
    fireEvent.click(subBtnEl);
    fireEvent.click(subBtnEl);
    expect(counterEl.textContent).toBe("20");

    fireEvent.change(inputEl, {
        target: {
            value: "5"
        }
    });
    fireEvent.click(addBtnEl);
    fireEvent.click(subBtnEl);
    fireEvent.click(subBtnEl);
    expect(counterEl.textContent).toBe("15");
});

test("counter contains correct className", () => {

    expect(counterEl.className).toBe("");
    
    fireEvent.change(inputEl, {
        target: {
            value: "50"
        }
    });
    fireEvent.click(addBtnEl);
    expect(counterEl.className).toBe("");
    fireEvent.click(addBtnEl);
    expect(counterEl.className).toBe("bg-success");
    fireEvent.click(addBtnEl);
    expect(counterEl.className).toBe("bg-success");

    fireEvent.click(subBtnEl);
    fireEvent.click(subBtnEl);
    expect(counterEl.className).toBe("");
    fireEvent.click(subBtnEl);
    fireEvent.click(subBtnEl);
    fireEvent.click(subBtnEl);
    expect(counterEl.className).toBe("bg-danger");

});

test('take the snap snap shot', () => {
  const tree = renderer
    .create(<Counter />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});