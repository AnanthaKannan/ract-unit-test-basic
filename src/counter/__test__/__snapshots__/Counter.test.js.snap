// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <h2
    data-testid="header"
  >
    My Counter
  </h2>
  <h3
    className=""
    data-testid="counter"
  >
    0
  </h3>
  <button
    data-testid="add-btn"
    onClick={[Function]}
  >
    +
  </button>
  <input
    className="App"
    data-testid="input"
    onChange={[Function]}
    type="number"
    value={1}
  />
  <button
    data-testid="sub-btn"
    onClick={[Function]}
  >
    -
  </button>
</div>
`;

exports[`take the snap snap shot 1`] = `
<div>
  <h2
    data-testid="header"
  >
    My Counter
  </h2>
  <h3
    className=""
    data-testid="counter"
  >
    0
  </h3>
  <button
    data-testid="add-btn"
    onClick={[Function]}
  >
    +
  </button>
  <input
    className="App"
    data-testid="input"
    onChange={[Function]}
    type="number"
    value={1}
  />
  <button
    data-testid="sub-btn"
    onClick={[Function]}
  >
    -
  </button>
</div>
`;
